app_dart/lib/src/request_handling/authentication.dart

```diff
@@ -135,68 +139,99 @@ class AuthenticationProvider {
     } else if (isCron) {
       // Authenticate cron requests that are not agents.
       return AuthenticatedContext._(clientContext: clientContext);
-    } else if (idToken != null) {
-      // Authenticate as a signed-in Google account via OAuth id token.
-      final HttpClient client = _httpClientProvider();
-      try {
-        final HttpClientRequest verifyTokenRequest = await client.getUrl(Uri.https(
-          'oauth2.googleapis.com',
-          '/tokeninfo',
-          <String, String>{
-            'id_token': idToken,
-          },
-        ));
-        final HttpClientResponse verifyTokenResponse = await verifyTokenRequest.close();
-
-        if (verifyTokenResponse.statusCode != HttpStatus.ok) {
-          final String body = await utf8.decodeStream(verifyTokenResponse);
-          log.warning('Token verification failed: ${verifyTokenResponse.statusCode}; $body');
-          throw const Unauthenticated('Invalid ID token');
-        }
-
-        final String tokenJson = await utf8.decodeStream(verifyTokenResponse);
-        TokenInfo token;
+    } else if (idTokenFromCookie != null || idTokenFromHeader != null) {
+      /// There are two possible sources for an id token:
+      ///
+      /// 1. Angular Dart app sends it as a Cookie
+      /// 2. Flutter app sends it as an HTTP header
+      ///
+      /// The Flutter application sends both since the client cookies are forwarded
+      /// in the request. One of these may be unauthenticated, so we need to try
+      /// authenticating both.
+      if (idTokenFromCookie != null) {
         try {
-          token = TokenInfo.fromJson(json.decode(tokenJson));
-        } on FormatException {
-          throw InternalServerError('Invalid JSON: "$tokenJson"');
-        }
+          return await authenticateIdToken(idTokenFromCookie,
+              clientContext: clientContext, log: log);
+        } catch (_) {}
+      }
 
-        final String clientId = await _config.oauthClientId;
-        assert(clientId != null);
-        if (token.audience != clientId) {
-          log.warning('Possible forged token: "${token.audience}" (expected "$clientId")');
-          throw const Unauthenticated('Invalid ID token');
-        }
+      if (idTokenFromHeader != null) {
+        return authenticateIdToken(idTokenFromHeader,
+            clientContext: clientContext, log: log);
+      }
```

Comment:
Why the above one has `await`, `try catch` but not this one?

Reply:
We wait to see if the user can be authenticated with the cookie id token first. This is to prevent making duplicate requests to the Google Auth API where we might be flagged for always sending two requests.

The try catch is to allow us fallback and try this method. If this last method doesn't authenticate, the user is not authenticated anyways.

Should I add a comment expanding on this more in the code?

Reply:
Yeah, a comment would be very helpful to clarify the differences.